packages/api/src/middleware/acl/hasFarmAccess.js

```diff
@@ -268,39 +269,42 @@ async function fromTaskManagementPlanAndLocation(req) {
   const farm_id = req.headers.farm_id;
   // harvest_tasks POST request body is an array
   const tasks = req.body.length ? req.body : [req.body];
-  const locationIds = new Set();
-  for (const { managementPlans, locations } of tasks) {
-    locations.forEach(({ location_id }) => locationIds.add(location_id));
-
-    for (const { planting_management_plan_id } of managementPlans || []) {
-      const managementPlan = await knex('management_plan')
-        .join(
-          'planting_management_plan',
-          'planting_management_plan.management_plan_id',
-          'management_plan.management_plan_id',
-        )
-        .join('crop_variety', 'crop_variety.crop_variety_id', 'management_plan.crop_variety_id')
-        .where('planting_management_plan.planting_management_plan_id', planting_management_plan_id)
-        .first();
-      if (managementPlan.farm_id !== farm_id) return {};
-
-      if (managementPlan.location_id) {
-        locationIds.add(managementPlan.location_id);
-      }
+  const { locationIds, plantingManagementPlanIds } = tasks.reduce(
+    ({ locationIds, plantingManagementPlanIds }, { locations, managementPlans }) => {
+      return {
+        locationIds: [...locationIds, ...(locations || []).map(({ location_id }) => location_id)],
+        plantingManagementPlanIds: [
+          ...plantingManagementPlanIds,
+          ...(managementPlans || []).map(
+            ({ planting_management_plan_id }) => planting_management_plan_id,
+          ),
+        ],
+      };
+    },
+    { locationIds: [], plantingManagementPlanIds: [] },
+  );
+  if (locationIds.length) {
+    const farmIds = await knex('location')
+      .whereIn('location_id', [...locationIds])
+      .pluck('farm_id');
+    if (farmIds.some((locationFarmId) => locationFarmId !== farm_id)) {
+      return {};
     }
   }
-  const farmIds = await knex('location')
-    .whereIn('location_id', [...locationIds])
-    .pluck('farm_id');
-
-  if (
-    farmIds.length !== locationIds.size || // check if all locationIds exist in the DB
-    new Set(farmIds).size !== 1 ||
-    farmIds[0] !== farm_id
-  ) {
-    return {};
+  if (plantingManagementPlanIds.length) {
+    const farmIds = await knex('planting_management_plan')
+      .join(
+        'management_plan',
+        'planting_management_plan.management_plan_id',
+        'management_plan.management_plan_id',
+      )
+      .join('crop_variety', 'crop_variety.crop_variety_id', 'management_plan.crop_variety_id')
+      .whereIn('planting_management_plan.planting_management_plan_id', plantingManagementPlanIds)
+      .pluck('farm_id');
+    if (farmIds.some((cropVarietyFarmId) => cropVarietyFarmId !== farm_id)) {
+      return {};
+    }
```

Comment:
@SayakaOno 
I think I remembered things wrong when I commented the failing harvest task test case.
Sorry about that.
Since planting_management_plan.location_id is already validated before it went into the DB, 
validating one of [crop_varity.farm_id, location.farm_id] should be enough

Reply:
No worries, I understand!
Thank you for taking care of this :)